### DISCUSSION TITLE
How to close an eframe::Frame or eframe::App?

### DISCUSSION THREAD

**Discussion by rokbok** [emilk/egui#4103] [Q&A]
It appears that there used to be a "close" function in  eframe::Frame, but that appears to be gone in 0.26.2. What is the correct way to ask for the window to be closed now? I might be missing something

**Comment by rokbok** [comment]
So I found  ui.ctx().send_viewport_cmd(ViewportCommand::Close); However, this doesn't really close the window. It remains open and visible for as long as the process continues to run.
EDIT: And running ctx.send_viewport_cmd(ViewportCommand::Visible(false)); in addition to Close does not work either. This seems to be related to #3655

**Comment by 39george** [comment]
I found that if you do this within `update` loop, it will block execution and freeze in the last `0.27.2` version: 
```rust
ctx.send_viewport_cmd(egui::ViewportCommand::Close);
```
But if you do this, it efficiently closes the window:
```rust
let ctx = ctx.clone();
std::thread::spawn(move || {
     ctx.send_viewport_cmd(egui::ViewportCommand::Close);
});
```
But this behaviour is not documented currently

**Reply by couleurm** [reply]
yeah i also need a way to run a command and close the window, it currently freezes whichever way i run them

will try this thread nesting out thanks

**Reply by thmxv** [reply]
I am running the last git version, not `0.27.2`, so this might be a bug that was corrected. But for me the `confirm_exit` example run fine, does close the window and does not freeze.

Edit: Apparently this behave a bit differently if the window is root (the first and main window) or not https://docs.rs/egui/latest/egui/struct.ViewportInfo.html#method.close_requested :

> If this is the root viewport, the application will exit after this frame unless you send a [crate::ViewportCommand::CancelClose](https://docs.rs/egui/latest/egui/viewport/enum.ViewportCommand.html#variant.CancelClose) command.
>
>If this is not the root viewport, it is up to the user to hide this viewport the next frame.

There is no documentation on how to do the hiding but there is another example `multiple_viewports` that does exactly that and that works here too (without threads)

**Reply by cyrilleberger** [reply]
Worked for me without the need for a thread on egui 0.32.
